internal/controller/clusterinstance_controller.go

```diff
@@ -253,120 +266,90 @@ func (r *ClusterInstanceReconciler) applyACMBackupLabelToInstallTemplates(
+// handleFinalizer ensures proper finalizer management for the ClusterInstance resource.
+// - Adds the finalizer if the object is not marked for deletion.
+// - Executes cleanup logic and removes the finalizer when the object is marked for deletion.
+// Returns a ctrl.Result to indicate whether the reconciler should requeue and an error if any operation fails.
+func (r *ClusterInstanceReconciler) handleFinalizer(
 	ctx context.Context,
 	log *zap.Logger,
 	clusterInstance *v1alpha1.ClusterInstance,
-) error {
+) (ctrl.Result, error) {
+	log = log.Named("handleFinalizer")
 
-	// Group the manifests by the sync-wave
-	// This is so that the manifests can be deleted in descending order of sync-wave
-	manifestGroups := map[int][]v1alpha1.ManifestReference{}
-	for _, manifest := range clusterInstance.Status.ManifestsRendered {
-		// check if the key exists in the map
-		if _, ok := manifestGroups[manifest.SyncWave]; !ok {
-			// if key doesn't exist, initialize the slice
-			manifestGroups[manifest.SyncWave] = make([]v1alpha1.ManifestReference, 0)
-		}
-		// append the value to the slice associated with the key
-		manifestGroups[manifest.SyncWave] = append(manifestGroups[manifest.SyncWave], manifest)
-	}
-
-	syncWaves := maps.Keys(manifestGroups)
-	// Sort the syncWaves in descending order
-	sort.Sort(sort.Reverse(sort.IntSlice(syncWaves)))
-
-	for _, syncWave := range syncWaves {
-		for _, manifest := range manifestGroups[syncWave] {
-			obj := &unstructured.Unstructured{}
-			obj.SetName(manifest.Name)
-			obj.SetNamespace(manifest.Namespace)
-			obj.SetAPIVersion(*manifest.APIGroup)
-			obj.SetKind(manifest.Kind)
-
-			if err := r.deleteResource(
-				ctx,
-				log,
-				ci.GenerateOwnedByLabelValue(clusterInstance.Namespace, clusterInstance.Name),
-				obj,
-			); err != nil {
-				return err
-			}
-		}
+	// Check if ClusterInstance is not being deleted
+	if clusterInstance.DeletionTimestamp.IsZero() {
+		return r.ensureFinalizer(ctx, log, clusterInstance)
 	}
-	log.Info("Successfully finalized ClusterInstance")
-	return nil
-}
 
-func (r *ClusterInstanceReconciler) deleteResource(
-	ctx context.Context,
-	log *zap.Logger,
-	owner string,
-	obj client.Object,
-) error {
+	// ClusterInstance is being deleted
+	if !controllerutil.ContainsFinalizer(clusterInstance, clusterInstanceFinalizer) {
+		// Finalizer already removed; no action needed
+		return ctrl.Result{}, nil
+	}
 
-	resourceId := ci.GetResourceId(
-		obj.GetObjectKind().GroupVersionKind().Kind, obj.GetNamespace(), obj.GetName(),
-	)
-	// Check that the manifest is logically owned-by the ClusterInstance
-	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(obj), obj); err != nil {
-		if errors.IsNotFound(err) {
-			log.Sugar().Infof("Skipping deletion of resource %s not found", resourceId)
-			return nil
+	log.Info("Running finalization logic for ClusterInstance")
+
+	// Create a DeletionContext
+	deletionCtx := deletion.NewDeletionContext(ctx, r.Client, log, nil)
```

Comment:
If this isn't actually a `context.Context` I'd pick a different name. Maybe `deletionInfo` or `deletionData`.

There is also the [value store associated with actual contexts](https://pkg.go.dev/context@go1.23.4#WithValue), but I'm not sure that's a good fit here as you're not actually transferring this across APIs.

Personally it doesn't seem crazy to pass a context, client, and log to the two other delete functions you have here and just get rid of the whole concept of a separate struct.

Reply:
Another option is to make this an actual long lived "handler" type struct where it is initialized with the client and log and passed into the reconciler at setup time, but I'll leave it up to you if you want to go that way.

Reply:
I personally prefer the Handler approach. Refactored accordingly.